### ISSUE TITLE
Hooks

### ISSUE THREAD

**Issue by wlandau** [ropensci/tarchetypes#44]
## Prework

* [x] I understand and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and  [contributing guidelines](https://github.com/ropensci/tarchetypes/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/ropensci/tarchetypes/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] New features take time and effort to create, and they take even more effort to maintain. So if the purpose of the feature is to resolve a struggle you are encountering personally, please consider first posting a "trouble" or "other" issue so we can discuss your use case and search for existing solutions first.

## Proposal

Automated hooks could let users prepend or wrap arbitrary R code around all the commands of all the targets in a target list. This could solve a multitude of issues, including:

* Custom prework, e.g. `conflicted` setup (cc @MilesMcBain)
* Restarts and timeouts.
* Custom encryption for all non-file targets.

## `tar_hook_before()`

```r
# Regular _targets.R
library(targets)
source("R/functions.R")
list(
  tar_target(x, {
    set_up_conflicted_package()
    do_work()
  }),
  tar_target(y, {
    set_up_conflicted_package()
    do_more_work()
  })
)
```

```r
# _targets.R with tar_hook_before()
library(targets)
library(magrittr)
source("R/functions.R")
targets <- list(
  tar_target(x, do_work()),
  tar_target(y, do_more_work())
) %>%
  tar_hook_before(set_up_conflicted_package)
```

## `tar_hook_inner()` and `tar_hook_outer()`

```r
# Regular _targets.R
library(targets)
source("R/functions.R")
list(
  tar_target(data, encrypt(get_data())),
  tar_target(analysis, encrypt(analyze(decrypt(data))))
)
```

```r
# _targets.R using inner and outer hooks
library(targets)
library(magrittr)
source("R/functions.R")
targets <- list(
  tar_target(data, get_data()),
  tar_target(analysis, analyze(data))
) %>%
  tar_hook_inner(encrypt(.x)) %>%
  tar_hook_outer(decrypt(.x))
```

**Comment by wlandau** [comment]
We could also use `tidyselect` to choose which targets to apply the hooks.

**Comment by wlandau** [comment]
Also, let's consider supporting arbitrary expressions instead of just single-argument functions. A hook could look something like `tar_hook_outer(target_object, hook = run_in_docker(code = .x, hostname = "localhost"))`, and the wrapped command could go in `.x`.

**Comment by wlandau** [comment]
A few implementation issues that will probably come up repeatedly in subsequent discussions:

### Composition of target lists

Should we return a flattened list of target objects, e.g. `unlist(list(targets), recursive = TRUE)`, or should we keep the target list nested in the way the user defines it originally, e.g. `tar_map(unlist = FALSE)`? I would think the latter. This outcome seems more likely to prevent surprises. Hooks should be minimally invasive when possible.

### Modification in place

Should the hooks modify the target objects in place, or should they return deep copies of those objects? If we were to make deep copies, we would probably need to flatten the target list. (Recursing over an arbitrarily nested list and making arbitrary changes requires passing by reference unless we come up with some messy index scheme to keep track of where we are in the list at all times.) Above, we already decided against flattening the target list, so we should go with in-place modifications.

### Return value

Given the decision to go with modification in place, we have a choice when it comes to the return value. We could either return the modified target list, or we could invisibly return `NULL`. I strongly prefer the latter because it emphasizes that target objects are modified in place, which can be counterintuitive.

### Drawback: potential to accidentally drop targets

Users might not read the docs and expect `tar_hook_before()` etc. to return target objects. In that case, those targets may not make it into the pipeline. This kind of mistake is not costly because no extra runtime is spent on misspecified targets. Inversely, returning the target list from the hooks could lead to incorrect target specification if the user forgets that the hooks modify things in place. So I think this is the right tradeoff.

### Drawback: metaprogramming limitations

Users might want to apply multiple hooks to a common template target object. However, this is already possible with `tar_map()`. The template can be created with `tar_target()` and possibly `tar_hook_first()` etc., and then `tar_map()` could be invoked afterwards.

### Best alternative

The best alternative I see to all this is to return a flattened list of deep-copied target objects. I still prefer not to do this because

1. It would be more invasive, and
2. Later on, if the demand is high enough, we can always build wrappers on top of `tar_hook_first()` etc. that make deep copies beforehand. We could either name those functions `tar_hook_first_copy()` etc. or think of another term besides "hook". But let's try to pursue slick workarounds before we jump to an implementation.

**Comment by wlandau** [comment]
TL;DR: I am trying to make hooks safe, even if that makes them a bit limited, at least at first.

**Comment by wlandau** [comment]
On second thought, I think it's fine if hooks flatten the target list. I don't think people need the actual structure in a target list all that often.

**Comment by wlandau** [comment]
So https://github.com/ropensci/tarchetypes/issues/44#issuecomment-831277768 is a justification for flattening the target list, not modification in place.

**Comment by wlandau** [comment]
Implemented with the deep copy route.
